sdk/core/Azure.Core.TestFramework/src/TestEnvironment.cs

```diff
@@ -175,6 +178,68 @@ public TokenCredential Credential
             }
         }
 
+        /// <summary>
+        /// Returns whether environment is ready to use. Should be overriden to provide service specific sampling scenario.
+        /// The test framework will wait until this returns true before starting tests.
+        /// Use this place to hook up logic that polls if eventual consistency has happened.
+        ///
+        /// Return true if environment is ready to use.
+        /// Return false if environment is not ready to use and framework should wait.
+        /// Throw if you want to fail the run fast.
+        /// </summary>
+        /// <returns>Whether environment is ready to use.</returns>
+        public virtual Task<bool> IsEnvironmentReady()
+        {
+            return Task.FromResult(true);
+        }
+
+        /// <summary>
+        /// Waits until environment becomes ready to use. See <see cref="IsEnvironmentReady"/> to define sampling scenario.
+        /// </summary>
+        /// <returns>A task.</returns>
+        public async Task WaitForEnvironment()
+        {
+            if (EnvironmentStateCache.TryGetValue(GetType(), out var environmentState))
+            {
+                switch (environmentState)
+                {
+                    case EnvironmentReadyState.Ready:
+                        return;
+                    case EnvironmentReadyState.Failed:
+                        throw new InvalidOperationException("The environment has not become ready, check your TestEnvironment.IsEnvironmentReady scenario.");
+                }
+            }
+
+            try
```

Comment:
Should we check the mode here?

Reply:
I was thinking of leaving that up to service owner to provide but for now we can do this until there's a need for customization.